cmd/ip.go

```diff
@@ -73,89 +69,89 @@ func (s ipManager) getAllIPs() []net.IP {
-	// ip.IsGlobalUnicast() returns true for:
-	// GUA = Global Unicast Address
-	// ULA = Unique Local Address
-	// We prefer GUA over ULA.
-	ipv6s := []net.IP{}
-	for _, ip := range ips {
-		if ip.IsGlobalUnicast() && ip.To4() == nil {
-			ipv6s = append(ipv6s, ip)
-		}
-	}
+func (s ipv6Stack) getBaseScore(ip net.IP) uint16 {
+	score := uint16(1)
 
-	return ipv6s
-}
+	if ip.To4() != nil {
+		return 0
+	}
+	if !ip.IsGlobalUnicast() {
+		return 0
+	}
 
-func (s ipv4Stack) filterIPs(ips []net.IP) []net.IP {
-	ipv4s := []net.IP{}
-	for _, ip := range ips {
-		if ip.To4() != nil {
-			ipv4s = append(ipv4s, ip)
+	if ipv6IsGUA(ip) {
+		score += 0x8
+		if ipv6IsEUI64(ip) {
+			score += 0x4
 		}
 	}
 
-	return ipv4s
+	return score
 }
 
-func (s ipv6Stack) sortIPs(ips []net.IP) []net.IP {
-	// Sort addresses placing GUAs first
-	sort.Slice(ips, func(i, j int) bool {
-		return ipv6IsGUA(ips[i]) && !ipv6IsGUA(ips[j]) ||
-			ipv6IsEUI64(ips[i]) && !ipv6IsEUI64(ips[j])
-	})
-	return ips
-}
+func (s ipv4Stack) getBaseScore(ip net.IP) uint16 {
+	score := uint16(1)
 
-func (s ipv4Stack) sortIPs(ips []net.IP) []net.IP {
-	// Sort addresses placing GUAs first, then Shared Address Space addresses,
-	// then private addresses, then loopback addresses.
-	sort.Slice(ips, func(i, j int) bool {
-		return ips[i].IsGlobalUnicast() && !ips[j].IsGlobalUnicast() ||
-			ipv4IsSAS(ips[i]) && !ipv4IsSAS(ips[j]) ||
-			ips[i].IsPrivate() && !ips[j].IsPrivate()
-	})
-	return ips
+	if ip.To4() == nil {
+		return 0
+	}
+	if ip.IsGlobalUnicast() {
+		score += 0x8
+	}
+	if ipv4IsSAS(ip) {
+		score += 0x4
+	}
+	if ip.IsPrivate() {
+		score += 0x2
+	}
+
+	return score
 }
 
-func (s ipManager) pickIP(ips []net.IP) net.IP {
-	netPrioritySubnets := []net.IPNet{}
-	for _, subnet := range s.cfg.prioritySubnets {
+// The score function evaluates the value of a given IP address and returns
+// a score of type uint64.
+// The higher the score, the more valuable the IP address.
+func (s ipManager) score(ip net.IP) uint64 {
+	// Score format:
+	// +------------+------------------+--------------+
+	// | Reserved   | Priority Subnets | Address Type |
+	// | (16 bits)  |    (32 bits)     |   (16 bits)  |
+	// +------------+------------------+--------------+
+	score := uint64(s.getBaseScore(ip))
+
+	// Scoring by priority subnet
+	for order, subnet := range s.cfg.prioritySubnets {
 		_, ipNet, err := net.ParseCIDR(subnet)
 		if err != nil {
 			log.WithError(err).WithField("subnet", subnet).Error("Couldn't parse subnet")
 			continue
 		}
-		netPrioritySubnets = append(netPrioritySubnets, *ipNet)
-	}
-
-	maxWeight := len(netPrioritySubnets)
-	weightedAddresses := make(map[string]int)
-	for _, ip := range ips {
-		weightedAddresses[ip.String()] = maxWeight
-		for i, ipNet := range netPrioritySubnets {
-			if ipNet.Contains(ip) {
-				weightedAddresses[ip.String()] = i
-				break
-			}
+		if ipNet.Contains(ip) {
+			score += uint64((1 + len(s.cfg.prioritySubnets) - order)) << 16
```

Comment:
I think the `+1` is not needed here anymore.

```suggestion
			score += uint64(len(s.cfg.prioritySubnets) - order) << 16
```